### ISSUE TITLE
error creating an ID-mapped copy of layer "..."

### ISSUE THREAD

**Issue by j0057** [containers/podman#8259]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When starting two rootful containers that share a base layer in a user namespace, I get an error `Error: error creating container storage: error creating an ID-mapped copy of layer "...": exit status 1: error during chown: error mapping host ID pair idtools.IDPair{UID:..., GID:...} for "..." to container: Host ID ... cannot be mapped to a container ID`.

I can reproduce it in  archlinux VMs using ext4 and btrfs, but not Fedora 33. Happens with `crun` as well as `runc`.

**Steps to reproduce the issue:**

1. As root: `podman run --rm -it --uidmap 0:10000:2000 --gidmap 0:10000:2000 --ip 10.88.0.10 --env 'discovery.type=single-node' docker://docker.io/library/elasticsearch:7.8.1`
2. As root: `podman run --rm -it --uidmap 0:12000:2000 --gidmap 0:12000:2000 --ip 10.88.0.11 --env 'SERVER_NAME=10.88.0.11' --env 'ELASTICSEARCH_HOSTS=["http://10.88.0.10:9200/"]' docker://docker.io/kibana:7.8.1`

**Describe the results you received:**

~`Error: error creating container storage: error creating an ID-mapped copy of layer "b97c3d2956a5832284e17580a4a34b8f12d1b8484f62400440b99ff1cb2c9198": exit status 1: error during chown: error mapping host ID pair idtools.IDPair{UID:111000, GID:111000} for "usr/share/terminfo/E/Eterm-color" to container: Host ID 111000 cannot be mapped to a container ID`~

`Error: error creating container storage: error creating an ID-mapped copy of layer "b97c3d2956a5832284e17580a4a34b8f12d1b8484f62400440b99ff1cb2c9198": exit status 1: error during chown: error mapping host ID pair idtools.IDPair{UID:12000, GID:12000} for "usr/share/terminfo/E/Eterm-color" to container: Host ID 12000 cannot be mapped to a container ID`

**Describe the results you expected:**

Two happily running containers.

**Additional information you deem important (e.g. issue happens only occasionally):**

The issue does not occur when using the same uidmap/gidmaps. The exact filename differs on each occasion. 

**Output of `podman version`:**

```
Version:      2.1.1
API Version:  2.0.0
Go Version:   go1.15.2
Git Commit:   9f6d6ba0b314d86521b66183c9ce48eaa2da1de2
Built:        Sat Sep 26 17:50:37 2020
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.1
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: Unknown
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: 35a2fa83022e56e18af7e6a865ba5d7165fa2a4a'
  cpus: 8
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: muon
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.9.3-arch1-1
  linkmode: dynamic
  memFree: 897466368
  memTotal: 33629425664
  ociRuntime:
    name: runc
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc92
      commit: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
      spec: 1.0.2-dev
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 34356064256
  swapTotal: 34357637120
  uptime: 66h 50m 13.88s (Approximately 2.75 days)
registries:
  search:
  - [redacted].azurecr.io
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 68
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 2.0.0
  Built: 1601135437
  BuiltTime: Sat Sep 26 17:50:37 2020
  GitCommit: 9f6d6ba0b314d86521b66183c9ce48eaa2da1de2
  GoVersion: go1.15.2
  OsArch: linux/amd64
  Version: 2.1.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Name            : podman
Version         : 2.1.1-1
Description     : Tool and library for running OCI-based containers in pods
Architecture    : x86_64
URL             : https://github.com/containers/libpod
Licenses        : Apache
Groups          : None
Provides        : None
Depends On      : cni-plugins  conmon  device-mapper  iptables  libseccomp  runc  slirp4netns  libsystemd  fuse-overlayfs  skopeo
Optional Deps   : podman-docker: for Docker-compatible CLI
                  btrfs-progs: support btrfs backend devices [installed]
                  catatonit: --init flag support
                  crun: support for unified cgroupsv2 [installed]
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 77.95 MiB
Packager        : Morten Linderud <foxboron@archlinux.org>
Build Date      : za 26 sep 2020 17:50:37 CEST
Install Date    : zo 11 okt 2020 10:55:53 CEST
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Desktop, home server, and VMs.

**Comment by giuseppe** [comment]
you are not assigning enough IDs to the user namespace: `Host ID 111000 cannot be mapped to a container ID`.

Can you try mapping that ID as well?

`--uidmap 0:10000:2000 --uidmap 111000:111000:1 --gidmap 0:10000:2000 --gidmap 111000:111000:1`

**Comment by j0057** [comment]
Hi Giuseppe, I'll investigate further next Monday.

**Comment by j0057** [comment]
@giuseppe, that's my mistake; I pasted the wrong error message. If I give the first container a uidmap/gidmap of `0:10000:10000`, and the second `0:20000:10000`, I get the error `Host ID 20000 cannot be mapped to a container ID`. I can see there are only files with uid 0 and 1000 within the container when I run the second container on Fedora.

**Comment by j0057** [comment]
Some more trial and error on my part:

- There's no difference when using cgroupsv1 or cgroupsv2
- It happens in archlinux VMs with btrfs, ext4 and xfs filesystems
- Setting `storage.driver = "overlay"` or `"btrfs"` in /etc/containers/storage.conf makes no difference
- Neither does setting `storage.options.mountopt = "nodev,metacopy=on"` instead of the default `"nodev"`

**Comment by j0057** [comment]
Interestingly, the error always seems to happen on files that are hardlinked to another; in the kibana container this is true for `/usr/share/terminfo/E/{Eterm,Eterm-color}` as well as `/usr/lib64/distutils/bcppcompiler.{pyc,pyo}`. It makes sense that if [chownByMapsMain][1] changed the owner via one of the two filenames, it would see the changed uid the second time and try to map that, which [toContainer][2] can't.

That doesn't yet explain why this does work on Fedora but not on Arch, is there some kernel option that allows for different owners of hardlinked files?

[1]: https://github.com/containers/storage/blob/9e2666784aaaa95b6ff6a6efc21404f0fd32c558/drivers/chown.go#L60
[2]: https://github.com/containers/storage/blob/53015455b60a41acdd8ab364a087d51c2b08b7b6/pkg/idtools/idtools.go#L104

**Comment by j0057** [comment]
Just confirmed that on the Fedora VM I can change the owner of one hardlink and leave the other untouched, and on the Arch VM this also changes both the files.

**Comment by j0057** [comment]
I've since learned that the owner of a file is recorded in the inode, so ordinarily two hardlinks cannot have different owners.

Following @giuseppe's hint in issue #5816 that he was able to reproduce that problem using `index=on`, and based on the hardlink behavior, as well as [the "Non-standard behavior" section in the overlayfs documentation][1], I added this to my `/etc/containers/storage.conf`:

```
[storage.options.overlay]
mountopts = "nodev,index=off"
```

And now I am able to start the two containers in my archlinux VM. Thanks for indulging my rubber duck debugging! :duck:

[1]: https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#non-standard-behavior
